personal_assistant/pa_contacts/views.py

```diff
@@ -1,12 +1,64 @@
 from django.shortcuts import render, redirect
+from django.utils import timezone
+from django.db.models.functions import ExtractMonth, ExtractDay
+from datetime import timedelta, date
+
 
 from .forms import ContactsForm
 from .models import Contact
 
 
 def main(request):
     contacts = Contact.objects.all()
-    return render(request, 'contacts/index.html', {'pa_contacts': contacts})
+    today = timezone.now().date()
+    days_param = request.GET.get('days', 7)
+
+    tabs = [
+        (7, 'Week'),
+        (30, 'Month'),
+        (90, '3 months'),
+    ]
+
+    try:
+        days = int(days_param)
+        if days <= 0:
+            days = 7
+    except ValueError:
+        days = 7
+
+    end_date = today + timezone.timedelta(days=days)
+
+    upcoming_birthdays_filtered = []
+
+    upcoming_birthdays = (
+        Contact.objects.annotate(
+            birth_month=ExtractMonth('birthday'),
+            birth_day=ExtractDay('birthday')
+        )
+    )
+
+    for contact in upcoming_birthdays:
+        if contact.birthday:
+            birth_month = contact.birth_month
+            birth_day = contact.birth_day
+
+            if (birth_month < today.month) or (birth_month == today.month and birth_day < today.day):
```

Comment:
Цей рядок не вміщується в 80-символьне обмеження по довжині рядка тож раджу його розбити на декілька:
```suggestion
            if (
                birth_month < today.month or
                (birth_month == today.month and birth_day < today.day)
            ):
```